src/pilot/assign_role.py

```diff
@@ -417,28 +439,29 @@ def define_storage_logical_disks(drac_client, raid_controller_name):
     # The ironic RAID 'physical_disks' property is optional. While it is
     # presently used by this script, it is envisioned that it will not
     # be in the future.
+    remaining_physical_disks = []
     if 'physical_disks' in os_logical_disk:
         os_physical_disk_names = os_logical_disk['physical_disks']
-        remaining_physical_disks = [d for d in raid_cntlr_physical_disks
-                                    if d.id not in os_physical_disk_names]
-    else:
-        remaining_physical_disks = raid_cntlr_physical_disks
+        for cntrl, disks in raid_cntlr_physical_disks.items():
+            _ = [remaining_physical_disks.append(disk) for disk in disks
+                 if disk.id not in os_physical_disk_names]
 
     # Define JBOD logical disks with the remaining physical disks.
     #
     # A successful call returns a list, which may be empty; otherwise,
     # None is returned.
-    jbod_capable = drac_client.is_jbod_capable(raid_controller_name)
-    jbod_logical_disks = define_jbod_logical_disks(
-        drac_client, remaining_physical_disks, raid_controller_name,
-        jbod_capable)
-
-    if jbod_logical_disks is None:
-        return None
-
     logical_disks = [os_logical_disk]
-    logical_disks.extend(jbod_logical_disks)
+    for raid_controller in raid_controllers:
+        jbod_capable = drac_client.is_jbod_capable(raid_controller)
+        jbod_logical_disks = define_jbod_logical_disks(
+            drac_client, remaining_physical_disks, raid_controller,
```

Comment:
I think there is a bug here.  We check to see if the RAID controller is JBOD capable, then call into define_jbod_logical_disks, but we pass in all of the remaining disks for all RAID controllers.  This means that we may treat some of the disks as being attached to a JBOD capable RAID controller when they are not, or vice versa.  I think we want to pass in the remaining disks for just the RAID controller specified.  Maybe a good way to do that would be to take the raid_cntlr_physical_disks dictionary and remove from it the disks we use for the RAID1.  Then it would have the remaining disks per RAID controller and it could be used here.

Reply:
Done